pkg/datastore/headers.go

```diff
@@ -0,0 +1,433 @@
+	if err := txn.SetEntry(badger.NewEntry(HeaderNumKey(h.Header.BlockNumber).Bytes(), slot).WithTTL(ttl)); err != nil {
+		return err
+	}
+
+	return txn.Commit()
+}
+
+func (s *Datastore) GetHeadersBySlot(ctx context.Context, slot uint64) ([]structs.HeaderAndTrace, error) {
+	el, _ := s.hc.GetContent(slot, slot, 1)
+	if el != nil {
+		return el, nil
+	}
+
+	data, err := s.TTLStorage.Get(ctx, HeaderKey(slot))
+	if err != nil && errors.Is(err, ds.ErrNotFound) {
+		return el, err
+	}
+
+	el = []structs.HeaderAndTrace{}
+	if err = json.Unmarshal(data, &el); err != nil {
+		return el, err
+	}
+
+	return el, err
+}
+
+func (s *Datastore) GetHeadersByBlockNum(ctx context.Context, blockNumber uint64) ([]structs.HeaderAndTrace, error) {
+	slot, err := s.TTLStorage.Get(ctx, HeaderNumKey(blockNumber))
+	if err != nil {
+		return nil, err
+	}
+	return s.GetHeadersBySlot(ctx, binary.LittleEndian.Uint64(slot))
+}
+
+func (s *Datastore) GetHeadersByBlockHash(ctx context.Context, hash types.Hash) ([]structs.HeaderAndTrace, error) {
+	slot, err := s.TTLStorage.Get(ctx, HeaderHashKey(hash))
+	if err != nil {
+		return nil, err
+	}
+	return s.GetHeadersBySlot(ctx, binary.LittleEndian.Uint64(slot))
+}
+
+func (s *Datastore) GetLatestHeaders(ctx context.Context, limit uint64) ([]structs.HeaderAndTrace, error) {
+	ls := s.hc.GetLatestSlot()
+	stop := ls - uint64(24*time.Hour/time.Second*12)
+	el, lastSlot := s.hc.GetContent(ls, stop, int(limit))
+
+	if el == nil {
+		el = []structs.HeaderAndTrace{}
+	}
+
+	// all from memory
+	if len(el) >= int(limit) {
+		return el[:limit], nil
+	}
+
+	initialSlot := lastSlot
+	readr := bytes.NewReader(nil)
+	dec := json.NewDecoder(readr)
+	for {
+		data, err := s.TTLStorage.Get(ctx, HeaderKey(initialSlot))
+		if err != nil {
+			if errors.Is(err, ds.ErrNotFound) {
+				return el, nil
+			}
+			return el, err
+		}
+		readr.Reset(data)
+		hnt := []structs.HeaderAndTrace{}
+		if err := dec.Decode(&hnt); err != nil {
+			return nil, err
+		}
+
+		el = append(el, hnt...)
+		initialSlot--
+		// introduce limit?
+	}
+}
+
+func (s *Datastore) deprecatedGetHeaders(ctx context.Context, key ds.Key) ([]structs.HeaderAndTrace, error) {
+	data, err := s.TTLStorage.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	return s.deprecatedUnsmarshalHeaders(data)
+}
+
+func (s *Datastore) deprecatedUnsmarshalHeaders(data []byte) ([]structs.HeaderAndTrace, error) {
+	var headers []structs.HeaderAndTrace
+	if err := json.Unmarshal(data, &headers); err != nil {
+		var header structs.HeaderAndTrace
+		if err := json.Unmarshal(data, &header); err != nil {
+			return nil, err
+		}
+		return []structs.HeaderAndTrace{header}, nil
+	}
+	return headers, nil
+}
+
+// SaveHeaders is meant to persist the all the keys under one key
+// As optimization in future this function can operate only on database, so instead from memory it may just reorganize keys
+func (s *Datastore) SaveHeaders(ctx context.Context, slots []uint64, ttl time.Duration) error {
```

Comment:
Why not `StoreHeaders` for consistency

Reply:
I wanted to distinct it a little bit from the store event. 
As this is meant for a scenario when go you from memory into the database.
It's not a big deal to rename it from Save but I think it would need to be something more descriptive than Store